### ISSUE TITLE
test_cached_decode_matches_full_forward fails on Apple Silicon GPU (11/12 runs): atol=2e-3 is calibrated for CUDA

### ISSUE THREAD

**Issue by ttchen8** [XHToken/Spark-MLX-LLM#4]
## Summary

`tests/test_model.py::test_cached_decode_matches_full_forward` fails on the Apple Silicon Metal backend in **11 out of 12 runs**. The same test passes 12/12 on CPU. The cause appears to be the tolerance rather than the model: `atol=2e-3` was calibrated against CUDA (per the comment in the test), while Metal's fused SDPA kernel diverges more between batched prefill and token-by-token decode.

The test also does not seed the RNG, so the model is randomly initialised on every run. That makes the failure look intermittent when it is actually near-deterministic.

## Reproduction

Environment:

- macOS 15 (Darwin 25.5.0), Apple M5 Pro, 24 GB unified memory
- Python 3.11.15, `mlx` 0.32.2, `mlx-lm` 0.31.3
- `spark-mlx-llm` 0.1.0 at commit `ef406ef`

```sh
$ python -m pytest tests/test_model.py -q
FAILED tests/test_model.py::test_cached_decode_matches_full_forward
1 failed, 4 passed in 0.91s
```

```
E       assert array(False, dtype=bool)
E        +  where ... = mx.allclose(..., rtol=0.002, atol=0.002)
tests/test_model.py:88: AssertionError
```

Forcing the CPU backend passes the whole suite:

```sh
$ python -c "
import mlx.core as mx; mx.set_default_device(mx.cpu)
import pytest, sys; sys.exit(pytest.main(['-q','tests/test_model.py']))"
5 passed in 0.59s
```

Failure rate over 12 consecutive runs of just this test:

| backend | failures |
|---|---|
| GPU (default on Apple Silicon) | **11 / 12** |
| CPU | 0 / 12 |

## Measured gap

Same test config (`make_args()`), prefill vs step-by-step decode, per seed:

| seed | GPU abs | GPU rel | CPU abs | CPU rel |
|---|---|---|---|---|
| 0 | 4.07e-3 | 1.21e-3 | 4.77e-7 | 1.41e-7 |
| 1 | 5.09e-3 | 1.20e-3 | 4.77e-7 | 1.13e-7 |
| 2 | 5.20e-3 | 1.40e-3 | 6.26e-7 | 1.68e-7 |
| 3 | 4.62e-3 | 1.45e-3 | 4.77e-7 | 1.50e-7 |
| 4 | 5.62e-3 | 1.87e-3 | 3.58e-7 | 1.19e-7 |
| 5 | 6.86e-3 | 1.81e-3 | 4.77e-7 | 1.26e-7 |

The GPU **absolute** difference (4.1e-3 – 6.9e-3) consistently exceeds `atol=2e-3`. Since `mx.allclose` uses `atol + rtol * |b|` and the logits here are O(1), the absolute term dominates for near-zero elements — which is why it fails so reliably.

## This is a backend trait, not a Spark issue

To isolate it I compared `mlx_lm.models.base.scaled_dot_product_attention` against a NumPy float64 reference, with no Spark code involved (causal mask, GQA 16/4, L=64, fp32 inputs):

| head_dim | CPU vs reference | GPU vs reference |
|---|---|---|
| 256 | 6.3e-8 | 8.5e-4 |
| 128 | 5.8e-8 | 5.0e-4 |
| 64 | 7.4e-8 | 4.9e-4 |

The ~5e-4 GPU divergence is present at every head_dim, so it is a property of MLX's fused Metal SDPA kernel (different reduction order between the prefill and single-token paths), not of this model. Argmax is unaffected — in my end-to-end runs on the real Spark-X2.5-4B checkpoint, top-1 next-token agreement between backends held across all prompts I tried.

## Suggested fix

Two independent changes, either of which helps:

**1. Seed the test** so failures are reproducible rather than intermittent:

```python
def test_cached_decode_matches_full_forward():
    mx.random.seed(0)
    model = Model(make_args())
```

**2. Widen the tolerance, or scale it per backend.** The existing comment already anticipates this for CUDA; Metal needs more headroom. Something like:

```python
# Fused attention kernels use different reduction orders between batched
# prefill and token-by-token decode. The gap is larger on Metal than CUDA.
tol = 2e-3 if mx.default_device() == mx.cpu else 1e-2
assert mx.allclose(full_logits, cached_logits, rtol=tol, atol=tol)
```

A stricter alternative that keeps the test sharp: compare argmax (which is what generation actually depends on) and keep a loose numeric check as a secondary assertion.

Given that Apple Silicon is the primary target for this package, a suite that fails out of the box there is likely to be the first thing new users hit — I did, before getting anything else working.

**Comment by simple-zheng** [comment]
Thanks for the detailed investigation and measurements.

This is addressed in #8.

The fix:

- makes model initialization deterministic with a fixed random seed;
- applies a Metal-specific absolute tolerance;
- retains the stricter tolerance on CPU and CUDA;
- verifies both argmax consistency and numerical closeness.

The updated test passed 20 consecutive runs on Apple Metal, passed on CPU, and the full test suite completed with `14 passed, 2 skipped`.

We will keep this issue open until the PR is merged.
